Cards/CardContainer.gd

```diff
@@ -10,93 +10,184 @@ class_name CardContainer
+			_draw_card()
 
 
 func discard_all_cards() -> void:
-	for card: CardWorld in cards:
-		card.queue_free()
-	cards.clear()
+	while cards_in_hand.size() > 0:
+		_discard_last_card()
 
 
-func remove_card(card: CardWorld) -> void:
-	var card_index: int = cards.find(card)
-	cards[card_index].queue_free()
-	cards.remove_at(card_index)
+func discard_random_card(amount: int) -> void:
+	var valid_cards_to_discard: Array[CardWorld] = []
+	
+	for card in cards_in_hand:
+		if card != queued_card:
+			valid_cards_to_discard.append(card)
+	
+	var amount_to_discard: int = mini(valid_cards_to_discard.size(), amount)
+	for discarding_index: int in amount_to_discard:
+		var random_index: int = Helpers.get_random_array_index(valid_cards_to_discard)
+		discard_card(valid_cards_to_discard[random_index])
+		valid_cards_to_discard.remove_at(random_index)
+
+
+func discard_card(card: CardWorld) -> void:
+	var card_index: int = cards_in_hand.find(card)
+	_discard_card_at_index(card_index)
+
+
+func get_draw_pile_size() -> int:
+	return draw_pile.size()
+
+
+func get_discard_pile_size() -> int:
+	return discard_pile.size()
+
+
+func _draw_card() -> void:
+	# if draw pile is empty, shuffle discard pile into draw pile. Clear discard pile.
+	if draw_pile.size() <= 0:
+		# every card is in our hand. Exit because there's nothing to draw
+		if discard_pile.size() <= 0:
+			return
+		discard_pile.shuffle()
+		draw_pile = discard_pile.duplicate()
+		discard_pile.clear()
+	
+	# top card on the draw pile
+	var drawn_card: CardBase = draw_pile[0]
+	draw_pile.remove_at(0)
+	
+	_create_card_in_world(drawn_card)
+	
+	on_card_counts_updated.emit()
+
+
+func _create_card_in_world(card_data: CardBase):
+	var card_instance: CardWorld = card_scene.instantiate()
+	add_child(card_instance)
+	cards_in_hand.append(card_instance)
+
+	card_instance.init_card(card_data)
+	
+	# bind mouse events
+	var card_click_handler: ClickHandler = card_instance.get_click_handler()
+	card_click_handler.on_click.connect(_on_card_clicked.bind(card_instance))
+	card_click_handler.on_mouse_hovering.connect(_on_card_hovering.bind(card_instance))
+	card_click_handler.on_unhover.connect(_on_card_unhovered.bind(card_instance))
+
+
+func _discard_card_at_index(card_index: int) -> void:
+	var card: CardWorld = cards_in_hand[card_index]
+	
+	# add to discard pile
+	discard_pile.append(card.card_data)
+	
+	# remove from world
+	cards_in_hand.remove_at(card_index)
+	card.queue_free()
+	
+	on_card_counts_updated.emit()
+
+
+func _discard_last_card() -> void:
+	if cards_in_hand.size() > 0:
+		_discard_card_at_index(cards_in_hand.size() - 1)
 
 
 func _on_phase_changed(new_phase: Enums.Phase, _old_phase: Enums.Phase) -> void:
 	if new_phase == Enums.Phase.PLAYER_ATTACKING:
-		deal_cards()
+		deal_to_hand_limit()
 	if new_phase == Enums.Phase.ENEMY_ATTACKING:
 		discard_all_cards()
 
 
 func _on_card_clicked(card: CardWorld) -> void:
-	if CardManager.is_card_queued():
+	if is_card_queued():
```

Comment:
will check this function again tomorrow with more sleep, because I feel like there is probably a simpler way to write that.

Reply:
Yeah, probably so. There's still some issues with card input as well if you spam click. I wanted to open a new issue related to this, where I also wanted to ask if we'd prefer a click or drag and drop input system.